tools/util/flags.cpp

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2023 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "flags.h"
+
+#include <cstring>
+#include <iostream>
+#include <regex>
+#include <string>
+#include <unordered_set>
+#include <variant>
+#include <vector>
+
+namespace flags {
+
+std::vector<std::string> positional_arguments;
+
+namespace {
+// Extracts the flag name from a potential token.
+// This function only looks for a '=', to split the flag name from the value for
+// long-form flags. Returns the name of the flag, prefixed with the hyphen(s).
+inline std::string get_flag_name(const std::string& flag, bool is_short_flag) {
+  if (is_short_flag) {
+    return flag;
+  }
+
+  size_t equal_index = flag.find('=');
+  if (equal_index == std::string::npos) {
+    return flag;
+  }
+  return flag.substr(0, equal_index);
+}
+
+// Parse a boolean flag. Returns `true` if the parsing succeeded, `false`
+// otherwise.
+bool parse_flag(Flag<bool>& flag, bool is_short_flag,
+                const std::string& token) {
+  if (is_short_flag) {
+    flag.value() = true;
+    return true;
+  }
+
+  const std::string raw_flag(token);
+  size_t equal_index = raw_flag.find('=');
+  if (equal_index == std::string::npos) {
+    flag.value() = true;
+    return true;
+  }
+
+  const std::string value = raw_flag.substr(equal_index + 1);
+  if (value == "true") {
+    flag.value() = true;
+    return true;
+  }
+
+  if (value == "false") {
+    flag.value() = false;
+    return true;
+  }
+
+  return false;
+}
+
+// Parse a string flag. Moved the iterator to the last flag's token if it's a
+// multi-token flag. Returns `true` if the parsing succeeded.
+bool parse_flag(Flag<std::string>& flag, bool is_short_flag,
+                const char*** iterator) {
+  const std::string raw_flag(**iterator);
+  const size_t equal_index = raw_flag.find('=');
+  if (is_short_flag || equal_index == std::string::npos) {
+    if ((*iterator)[1] == nullptr) {
+      return false;
+    }
+
+    flag.value() = (*iterator)[1];
+    *iterator += 1;
+    return true;
+  }
+
+  const std::string value = raw_flag.substr(equal_index + 1);
+  flag.value() = value;
+  return true;
```

Comment:
Why don't you update iterator in this case?

Reply:
The iterator is always moving forward in the main parsing logic:
 - move iterator by 1
 - call parse.
 
 This works fine for flags like `-g` or `--foo=bar`.
 The issue is with flags like `--foo bar`, suddenly, a single flag is not made of one, but 2 tokens.
 This means we need to move the iterator by one additional step, but only in that case.
 This logic depending on the flag type, and 1st token content, it's handled down at the per-type flag parsing logic.
 (Since boolean cannot have a split `-g true` version).
 
 Now why can't we have a -g true, that's to avoid ambiguities:
  - we are used to use `-g` to say "enable the -g option".
  - using -g false means -g can take no, or maybe one parameter, and it depends on the value after.
    With only flags, it's easy: does the next option starts with a `-`? no? then it's a split boolean flag.
    But it is not possible since we also have positional arguments.   
 - `spirv-foo -g my-file` ->  my-file is not a badly written "true" or "false", but positional.
 
 This constraint means the logic to move iterators forward is not the same for all flag types. Adding this logic on the high level loop would mean duplicating it:
  - move token by one
  - parse token -> determine flag type -> run per-type parsing logic
  - determine flag type -> maybe move token by 1 again
  
  So I thought it would be better to pass the iterator, with the contract that the iterator will end on on the last parsed token. Hence the +1 only when we have a double-token flag.
  
  Will add a comment.